### ISSUE TITLE
Control actions and highlighted state on layouts

### ISSUE THREAD

**Issue by parski** [linkedin/LayoutKit#77]
My team and I are trying out this library for a feed we are incorporating into our app and we really like LayoutKit so far. We would like to thank you for open sourcing this and we find that it's a really great way to build performant user interfaces. However, we are really missing interactivity on layouts that are not `ButtonLayout`s. We want to be able to control the highlighted state of any layout.

Here's an example use case we stumbled upon that caused our desire for this feature:

We have a feed with cells laid out vertically stacked. These cells have different layouts, one of which contains an `ImageLayout` as a sublayout. We are able to feign the highlighted state on the rest of the layout by making it transparent and making the views in a class extending `UICollectionViewCell` that has a highlighted state. Of course this does not pertain to the layout containing an image and it's a bit of a hack.

Is there a recommended practice for how to achieve this already using LayoutKit? We are able to identify the sublayouts of a given layout but the views they wrap are inaccessible to us.

Should there not be a recommended practice to achieve this kind of behavior we propose introducing two new protocols to LayoutKit, `Actionable` and `Highlightable`.

`Actionable` would allow the layout to respond to presses with an action and during the press highlight all children conforming to `Highlightable` until a descendant conforms to `Actionable`.

`Highlightable` allows a view/layout to assume a highlighted state, altering it's appearance without changing the layout, when passed an action from the closest parent conforming to `Actionable`. The properties of this highlighted state would preferably be specified in the `config` closure.

I should also add that we are willing to contribute to this project should you accept this proposal.

**Comment by AlexLittlejohn** [comment]
Perhaps you can just store the highlighted state and then perform a layout update on user interaction? Updating the layout is very efficient especially if you make use of `viewReuseId`.

**Comment by nicksnyder** [comment]
Please read the discussions #21 and #75.

> We want to be able to control the highlighted state of any layout.

To the extent that you wish to make style changes to views that do not impact layout (i.e. size or location), you can just capture references to these views from the `config` block (be mindful of retain cycles) and update your style properties whenever you want.

Alternatively, you can do as @AlexLittlejohn said and just relayout with a highlighted state.

> We are able to identify the sublayouts of a given layout but the views they wrap are inaccessible to us.

I don't quite understand what you mean by this or why it is a problem. Can you elaborate? Any view you create with LayoutKit can be captured in the `config` block.

> we propose introducing two new protocols to LayoutKit, Actionable and Highlightable

These protocols sound like something a UIView should implement, not a Layout. Layouts have a single responsibility: computing the frame. Styling and interaction are out of scope and generally handled in the config block or by the UIView produced by the Layout.

**Comment by nicksnyder** [comment]
@parski ping, did my suggestions help?

**Comment by parski** [comment]
Excuse my slow response.

Your suggestion certainly helped. We solved this by using a custom view to be wrapped by our layouts and letting that view handle it's states and control actions. After some discussion we concluded that, like you said, it should be our `UIView` subclasses that handle the states and actions, not our layouts, so we withdraw our proposal and consider this issue closed.

Again, we are very satisfied with LayoutKit. Great work! ❤️
